

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>R Code Example: Gene Set Analysis and Binarization (hdWGCNA integration) &mdash; FOX 0.0.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=d45e8c67"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Python Example: Running the Hypergeometric Test" href="intro.html" />
    <link rel="prev" title="FOX package" href="FOX.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            FOX
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="FOX.html">FOX package</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">R Code Example: Gene Set Analysis and Binarization (hdWGCNA integration)</a></li>
<li class="toctree-l1"><a class="reference internal" href="intro.html">Python Example: Running the Hypergeometric Test</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">FOX</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">R Code Example: Gene Set Analysis and Binarization (hdWGCNA integration)</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/hdWGCNA_integ.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="r-code-example-gene-set-analysis-and-binarization-hdwgcna-integration">
<h1>R Code Example: Gene Set Analysis and Binarization (hdWGCNA integration)<a class="headerlink" href="#r-code-example-gene-set-analysis-and-binarization-hdwgcna-integration" title="Link to this heading"></a></h1>
<p>This section demonstrates how to perform gene set analysis using R, binarize clusters, and compute Jensen-Shannon Divergence (JSD) for comparing modules and clusters in gene expression data.</p>
<p>hub_df was generated from following the tutorial here: <a class="reference external" href="https://smorabit.github.io/hdWGCNA/articles/basic_tutorial.html">hdWGCNA</a>. Scroll down to section called “Getting the module assignment table”.</p>
<p>Step 1: Create Gene Sets for Each Module</p>
<p>First, we group the data by modules and create gene sets for each module.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># module is your regulon</span>
<span class="n">df_grouped</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">hub_df</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">group_by</span><span class="p">(</span><span class="n">module</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">summarize</span><span class="p">(</span><span class="n">genes_in_module</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">gene_name</span><span class="p">),</span><span class="w"> </span><span class="n">.groups</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;drop&quot;</span><span class="p">)</span>

<span class="c1"># Create gene sets for each module</span>
<span class="n">all.sets</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">lapply</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="nf">nrow</span><span class="p">(</span><span class="n">df_grouped</span><span class="p">),</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">GeneSet</span><span class="p">(</span><span class="n">df_grouped</span><span class="o">$</span><span class="n">genes_in_module</span><span class="p">[[</span><span class="n">i</span><span class="p">]],</span><span class="w"> </span><span class="n">setName</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">df_grouped</span><span class="o">$</span><span class="n">module</span><span class="p">[</span><span class="n">i</span><span class="p">]))</span>
<span class="p">})</span>

<span class="c1"># View the result</span>
<span class="nf">print</span><span class="p">(</span><span class="n">all.sets</span><span class="p">)</span>
<span class="n">all.sets</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">GeneSetCollection</span><span class="p">(</span><span class="n">all.sets</span><span class="p">)</span>
</pre></div>
</div>
<p>Step 2: Prepare Expression Matrix
Next, we extract the expression matrix from a Seurat object, which contains gene expression data.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">exprMatrx</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">seurat_obj</span><span class="p">[[</span><span class="s">&quot;RNA&quot;</span><span class="p">]]</span><span class="o">$</span><span class="n">counts</span>
</pre></div>
</div>
<p>Step 3: Calculate AUC and Assign Thresholds
Using the <cite>AUCell</cite> package, we calculate the Area Under the Curve (AUC) for the gene sets and explore the thresholds.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">cells_AUC</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">AUCell_calcAUC</span><span class="p">(</span><span class="n">geneSets</span><span class="p">,</span><span class="w"> </span><span class="n">cells_rankings</span><span class="p">)</span>
<span class="n">cells_assignment</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">AUCell_exploreThresholds</span><span class="p">(</span><span class="n">cells_AUC</span><span class="p">,</span><span class="w"> </span><span class="n">plotHist</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">nCores</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">assign</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<p>Step 4: Binarize Clusters Based on Conditions
We binarize clusters in the Seurat object based on cell type or condition.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># Loop through each cell type to binarize clusters (remember, pick your column)</span>
<span class="n">binarize_clusters</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">lapply</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">seurat_obj</span><span class="o">@</span><span class="n">meta.data</span><span class="o">$</span><span class="n">cell_type</span><span class="p">)),</span>
<span class="w">    </span><span class="kr">function</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="n">cluster</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">ifelse</span><span class="p">(</span><span class="n">seurat_obj</span><span class="o">@</span><span class="n">meta.data</span><span class="o">$</span><span class="n">cell_type</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)</span>
<span class="w">        </span><span class="kr">return</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span>
<span class="p">})</span>
<span class="nf">names</span><span class="p">(</span><span class="n">binarize_clusters</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">unique</span><span class="p">(</span><span class="n">seurat_obj</span><span class="o">@</span><span class="n">meta.data</span><span class="o">$</span><span class="n">cell_type</span><span class="p">)</span>
</pre></div>
</div>
<p>Step 5: Binarize Regulons Based on AUC Thresholds
Now, we binarize the regulons (gene modules) based on their AUC values and the selected thresholds.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">binarize_regulons</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">lapply</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">cells_AUC</span><span class="p">)),</span>
<span class="w">    </span><span class="kr">function</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="n">threshold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="n">cells_assignment</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span><span class="o">$</span><span class="n">aucThr</span><span class="o">$</span><span class="n">selected</span><span class="p">)</span>
<span class="w">        </span><span class="n">b_values</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ifelse</span><span class="p">(</span><span class="n">cells_AUC</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">threshold</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)</span>
<span class="w">        </span><span class="kr">return</span><span class="p">(</span><span class="n">b_values</span><span class="p">)</span>
<span class="p">})</span>
<span class="nf">names</span><span class="p">(</span><span class="n">binarize_regulons</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">cells_AUC</span><span class="p">)</span>
</pre></div>
</div>
<p>Step 6: Compute Jensen-Shannon Divergence (JSD)
Finally, we compute the Jensen-Shannon Divergence between the binarized clusters and regulons using the <a class="reference external" href="https://cran.r-project.org/web/packages/philentropy/index.html">philentropy</a> package.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="w"> </span><span class="nf">library</span><span class="p">(</span><span class="n">philentropy</span><span class="p">)</span>

<span class="w"> </span><span class="c1"># Initialize a matrix to store the JSD values</span>
<span class="w"> </span><span class="n">jsd_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">matrix</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">binarize_clusters</span><span class="p">),</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">binarize_regulons</span><span class="p">))</span>

<span class="w"> </span><span class="c1"># Set row and column names of the matrix</span>
<span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">jsd_matrix</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">binarize_clusters</span><span class="p">)</span>
<span class="w"> </span><span class="nf">colnames</span><span class="p">(</span><span class="n">jsd_matrix</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">binarize_regulons</span><span class="p">)</span>

<span class="w"> </span><span class="c1"># Loop through each combination of binarize_clusters and binarize_regulons</span>
<span class="w"> </span><span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">binarize_clusters</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">   </span><span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">i2</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">binarize_regulons</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">     </span><span class="n">t</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">binarize_clusters</span><span class="p">[[</span><span class="n">i</span><span class="p">]])</span><span class="w">  </span><span class="c1"># Length of the current vector</span>
<span class="w">     </span><span class="n">x_count</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rbind</span><span class="p">(</span><span class="n">binarize_clusters</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">t</span><span class="p">,</span><span class="w"> </span><span class="nf">as.double</span><span class="p">(</span><span class="n">binarize_regulons</span><span class="p">[[</span><span class="n">i2</span><span class="p">]])</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">t</span><span class="p">)</span><span class="w">  </span><span class="c1"># Normalize the vectors</span>
<span class="w">     </span><span class="n">jsd_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="n">i2</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">(</span><span class="nf">JSD</span><span class="p">(</span><span class="n">x_count</span><span class="p">,</span><span class="w"> </span><span class="n">unit</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;log2&quot;</span><span class="p">))</span>
<span class="w">   </span><span class="p">}</span>
<span class="w"> </span><span class="p">}</span>
<span class="nf">write.csv</span><span class="p">(</span><span class="n">jsd_matrix</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;RSS_scores.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Step 7: Plug into FOX.</p>
<p>Take the cells_AUC and jsd_matrix and export them to .csv files. Then plug them into FOX with your desired cluster names.</p>
<blockquote>
<div><p>Conclusion</p>
</div></blockquote>
<p>This analysis demonstrates the process of binarizing clusters and regulons, followed by computing Jensen-Shannon Divergence to compare gene sets. These steps are useful for understanding the relationships between gene modules and cell clusters, especially in single-cell RNA-seq data. This can be applied to a variety of different gene network libraries.</p>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="FOX.html" class="btn btn-neutral float-left" title="FOX package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="intro.html" class="btn btn-neutral float-right" title="Python Example: Running the Hypergeometric Test" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Moses Apostol.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>